core/ir/riscv64/codec.c

```diff
@@ -0,0 +1,1413 @@
+    int nsrc = 0, ndst = 0;
+    byte *next_pc;
+
+    CLIENT_ASSERT(instr->opcode == OP_INVALID || instr->opcode == OP_UNDECODED,
+                  "decode: instr is already decoded, may need to call instr_reset()");
+
+    switch (width) {
+    case 4:
+        inst |= *((uint16_t *)pc + 1) << 16;
+        info = get_rv_instr_info(inst, instr_infos_trie);
+        break;
+    case 2: info = get_rvc_instr_info(inst, 64); break;
+    default:
+        LOG(THREAD, LOG_INTERP, 3, "decode: unhandled instruction width %d at " PFX "\n",
+            width, pc);
+        CLIENT_ASSERT(false, "decode: invalid instr width");
+        return NULL;
+    }
+    next_pc = pc + width;
+
+    if (info == NULL) {
+        LOG(THREAD, LOG_INTERP, 3, "decode: unknown instruction 0x%08x at " PFX "\n",
+            inst, pc);
+        return NULL;
+    }
+
+    nsrc = INFO_NSRC(info->info.opcode);
+    ndst = INFO_NDST(info->info.opcode);
+    CLIENT_ASSERT(ndst >= 0 || ndst <= 1, "Invalid number of destination operands.");
+    CLIENT_ASSERT(nsrc >= 0 || nsrc <= 4, "Invalid number of source operands.");
+
+    instr_set_opcode(instr, info->info.type);
+    instr_set_num_opnds(dcontext, instr, ndst, nsrc);
+
+    CLIENT_ASSERT(info->info.dst1_type < RISCV64_FLD_CNT, "Invalid dst1_type.");
+    if (ndst > 0 &&
+        !opnd_decoders[info->info.dst1_type](dcontext, inst, info->info.dst1_size, pc,
+                                             orig_pc, 0, instr))
+        goto decode_failure;
+    switch (nsrc) {
+    case 4:
+        CLIENT_ASSERT(info->info.dst2_type < RISCV64_FLD_CNT, "Invalid dst2_type.");
+        if (!opnd_decoders[info->info.dst2_type](dcontext, inst, info->info.dst2_size, pc,
+                                                 orig_pc, 3, instr))
+            goto decode_failure;
+    case 3:
+        CLIENT_ASSERT(info->info.src3_type < RISCV64_FLD_CNT, "Invalid src3_type.");
+        if (!opnd_decoders[info->info.src3_type](dcontext, inst, info->info.src3_size, pc,
+                                                 orig_pc, 2, instr))
+            goto decode_failure;
+    case 2:
+        CLIENT_ASSERT(info->info.src2_type < RISCV64_FLD_CNT, "Invalid src2_type.");
+        if (!opnd_decoders[info->info.src2_type](dcontext, inst, info->info.src2_size, pc,
+                                                 orig_pc, 1, instr))
+            goto decode_failure;
+    case 1:
+        CLIENT_ASSERT(info->info.src1_type < RISCV64_FLD_CNT, "Invalid src1_type.");
+        if (!opnd_decoders[info->info.src1_type](dcontext, inst, info->info.src1_size, pc,
+                                                 orig_pc, 0, instr))
+            goto decode_failure;
+    case 0: break;
+    default: ASSERT_NOT_REACHED();
+    }
+
+    if (orig_pc != pc) {
+        /* We do not want to copy when encoding and condone an invalid
+         * relative target.
+         * FIXME i#3544: Add re-relativization support without having to re-encode.
```

Comment:
This will cause re-encoding of every single instr...the FIXME is already there so no suggestion just talking out loud.

Reply:
Yeah, that will be painful but I don't feel confident enough to tackle this issue ar this point.